lib/AST/ASTImporter.cpp

```diff
@@ -3294,25 +3326,37 @@ ExpectedDecl ASTNodeImporter::VisitFriendDecl(FriendDecl *D) {
 
   // Determine whether we've already imported this decl.
   // FriendDecl is not a NamedDecl so we cannot use lookup.
+  // We try to maintain order and count of redundant friend declarations.
   auto *RD = cast<CXXRecordDecl>(DC);
   FriendDecl *ImportedFriend = RD->getFirstFriend();
+  SmallVector<FriendDecl *, 2> ImportedFriends;
 
   while (ImportedFriend) {
+    // FIXME: Can we not use isStructuralMatch only?
```

Comment:
I think we could. We should have an overload of `isStructuralMatch(QualType, QualType)`. The check for already imported Types in `IsStructurallyEquivalent` is redundant anyway.

Reply:
But this should be in a separate PR.

Reply:
I think it is better to leave this as is:
The `ASTImporter::IsStructurallyEquivalent` with `QualType` does a `hasSameType` call if the type was already imported (no structural eq check, maybe it is more efficient).